pkg/kubecrypto/certs.go

```diff
@@ -0,0 +1,245 @@
+package kubecrypto
+
+import (
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	ocrypto "github.com/scylladb/scylla-operator/pkg/crypto"
+	"github.com/scylladb/scylla-operator/pkg/helpers"
+	"github.com/scylladb/scylla-operator/pkg/naming"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+)
+
+func needsRefresh(existingCert *x509.Certificate, now time.Time, refresh time.Duration, desiredCert *x509.Certificate, issuerPublicKey *rsa.PublicKey) string {
+	// Don't check notBefore to avoid issues on time skew.
+	// notAfter is fine as the cert should never be close to it.
+	if now.After(existingCert.NotAfter) {
+		return "already expired"
+	}
+
+	validity := existingCert.NotAfter.Sub(existingCert.NotBefore)
+	at80Percent := existingCert.NotAfter.Add(-validity / 5)
+	if now.After(at80Percent) {
+		return fmt.Sprintf("past its latest possible refresh time %v", at80Percent)
+	}
+
+	refreshDate := existingCert.NotBefore.Add(refresh)
+	if now.After(refreshDate) {
+		return fmt.Sprintf("past its refresh time %v", refreshDate)
+	}
+
+	existingCertTemplate := ocrypto.ExtractDesiredFieldsFromTemplate(existingCert)
+	desiredCertTemplate := ocrypto.ExtractDesiredFieldsFromTemplate(desiredCert)
+	if !reflect.DeepEqual(desiredCertTemplate, existingCertTemplate) {
+		klog.V(2).InfoS(
```

Comment:
TODO (@tnozicka): think about adding the object key so it's easy to grep

Reply:
done in https://github.com/scylladb/scylla-operator/compare/063205ff4e8e5b382505bf9352908bf5f4c946c1..cc09393ae1877aa19faab1d2d7236034b7690809